### PR TITLE
Better handle device init for Pointing Devices

### PR DESCRIPTION
**Pull request by drashna**
## Description

Noticed when messing with the code, that bad/poor initialization of pointing device sensors wasn't really handled gracefully. 

Changes and adds some logic to disable querying a device if it doesn't init properly.

Unfortunately, not all devices have a way to check if it's been initialized properly, and some could use an overhaul for the logic.  But for now, just allow for basic logic to be used. 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] Enhancement/optimization

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).
  - Code compiles for all sensors (and includes some additional fixes to do so), and pmw3360 sensor works on ploopy mouse.

### CODE REVIEW

**Review by KarlK90 — CHANGES_REQUESTED** [review]
Hi @drashna, I have some suggestions for the implementation. I would move the init checking logic out of the high level driver in `pointing_device.c` into the lower level drivers, so the check is not done by the caller but by the callee.

That would require the driver to become stateful like a `pointing_device_driver.state` field instead of a single global boolean.

```c
typedef enum {
    PTD_UNINITIALIZED,
    PTD_INITIALIZED,
    PTD_ERROR,
} pointing_device_driver_state_t;

const pointing_device_driver_t pointing_device_driver = {
    .init       = init,
    .state      = PTD_UNINITIALIZED,
    .get_report = adns9800_get_report_driver,
    .set_cpi    = adns9800_set_cpi,
    .get_cpi    = adns9800_get_cpi
};

```

```c
pointing_device_driver_state_t cirque_pinnacle_init(void) {
    state = PTD_INITIALIZED; // all ok
    state = PTD_ERROR; // something went wrong
    return state;
}
```

All driver provided functions do now check themselves if they have been properly initialized:

```c
report_mouse_t cirque_pinnacle_get_report(report_mouse_t mouse_report) {
    if(pointing_device_driver.state != PTD_INITIALIZED){
        return mouse_report;
    }
}
```

also the `get_cpi` and `set_cpi` functions, where appropriate, would follow the same pattern e.g.:

```c
#include "pointing_device.h"
extern pointing_device_driver_t pointing_device_driver;

uint16_t pmw3360_get_cpi(void) {
    if(pointing_device_driver.state != PTD_INITIALIZED){
        return 0; // zero is a good error value or rather 65536?
    }

    uint8_t cpival = pmw3360_read(REG_Config1);
    return (uint16_t)((cpival + 1) & 0xFF) * CPI_STEP;
}

void pmw3360_set_cpi(uint16_t cpi) {
    if(pointing_device_driver.state != PTD_INITIALIZED){
        return;
    }

    uint8_t cpival = constrain((cpi / CPI_STEP) - 1, 0, MAX_CPI);
    pmw3360_write(REG_Config1, cpival);
}
```

In my opinion this change is necessary because these functions are a "public" API that can be called on a user/keyboard level and these should not be bothered with doing the checks themselves. Also the global boolean `pointing_device_initialized` moves away from the `pointing_device_driver_t` bundeling.

One hard compile time error is currently in the code when setting the pointing device driver to `custom` because `pointing_device_driver_init()` still returns `void`.

```
Compiling: quantum/pointing_device_drivers.c                                                       quantum/pointing_device_drivers.c:303:19: error: initialization of '_Bool (*)(void)' from incompatible pointer type 'void (*)(void)' [-Werror=incompatible-pointer-types]
     .init       = pointing_device_driver_init,
```

Let me know what you think :slightly_smiling_face:

**Comment by KarlK90** [comment]
Updated the example code to use an enum which is cleaner

File: quantum/pointing_device_drivers.c [hunk 1]
```diff
@@ -44,13 +49,17 @@ report_mouse_t adns5050_get_report(report_mouse_t mouse_report) {
 
 // clang-format off
 const pointing_device_driver_t pointing_device_driver = {
-    .init         = adns5050_init,
+    .init         = init,
```
Comment (daskygit) [inline quantum/pointing_device_drivers.c:52]: Maybe these should be renamed as to match of the others `devicename_init`.

**Comment by drashna** [comment]
@KarlK90 that was on my mind, I just wasn't sure about the implementation.  But thanks for the good suggestions there

**Comment by daskygit** [comment]
Also how basic were you planning to keep this PR? You can do a checksum verification of the firmware upload on ADNS9800 and maybe the pimoroni should also be reworked as it has it's own failure mode after X amount of bad reads. (although didn't take into consideration stopping writes)

**Comment by drashna** [comment]
I would like to keep it basic and then work on building it up later.

for the adns9800, it doesn't have some of the implementation to handle that, so would need to actually build that out.  But I did want to do that, as well.

**Comment by KarlK90** [comment]
> @KarlK90 that was on my mind, I just wasn't sure about the implementation. But thanks for the good suggestions there

Great, you are welcome! :slightly_smiling_face:

**Pull request closed without merging**